### PR TITLE
Temp keys user policies

### PR DESCRIPTION
**Pull request by chenriksson**
Adding support for user security policies, including policies for use of temp keys and min client version

Onboarding and auditing will be separate PRs

### CODE REVIEW

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 1]
```diff
@@ -293,6 +294,7 @@ private HttpStatusCodeWithBodyResult VerifyPackageKeyInternal(User user, Credent
         [RequireSsl]
         [ApiAuthorize]
         [ApiScopeRequired(NuGetScopes.PackagePush, NuGetScopes.PackagePushVersion)]
+        [SecurityPolicy(SecurityPolicyAction.PackagePush)]
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: consider merging SecurityPolicy attribute with ApiAuthorize attribute.

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 2]
```diff
@@ -201,6 +202,11 @@ protected override void Load(ContainerBuilder builder)
                 .As<IStatusService>()
                 .InstancePerLifetimeScope();
 
+            builder.RegisterType<SecurityPolicyService>()
+                .AsSelf()
+                .As<ISecurityPolicyService>()
+                .InstancePerLifetimeScope();
```
Comment (skofman1) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: Any reason for ISecurityPolicyService not to be a single instance?

File: src/NuGetGallery/Security/SecurityPolicyService.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+using NuGetGallery.Filters;
+
+namespace NuGetGallery.Security
+{
+    /// <summary>
+    /// Service that looks up and evaluates security policies for attributed controller actions.
+    /// </summary>
+    public class SecurityPolicyService : ISecurityPolicyService
+    {
+        public IEnumerable<UserSecurityPolicyHandler> UserPolicyHandlers { get; private set; }
```
Comment (skofman1) [inline src/NuGetGallery/Security/SecurityPolicyService.cs]: consider making this a static

**Comment by skofman1** [comment]
> // Copyright (c) .NET Foundation. All rights reserved.

null checks

---
Refers to: src/NuGetGallery/Security/UserSecurityPolicyContext.cs:1 in d018112. [](commit_id = d0181124ef39c44b94caf8df4d2982d1aef14996, deletion_comment = False)

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs [hunk 4]
```diff
@@ -119,6 +119,18 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
...
+                .HasKey(c => c.Key);
+
+            modelBuilder.Entity<UserSecurityPolicy>()
+                .HasRequired<User>(p => p.User)
+                .WithMany(cr => cr.SecurityPolicies)
+                .HasForeignKey(p => p.UserKey)
+                .WillCascadeOnDelete(true);
+
+            modelBuilder.Entity<UserSecurityPolicy>()
+                .HasKey(p => p.Key);
```
Comment (cristinamanum) [inline src/NuGetGallery.Core/Entities/EntitiesContext.cs]: Is not this duplication of the line 123.

**Comment by skofman1** [comment]
Looks great! 👍

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        {
+        }
+
+        /// <summary>
+        /// In case of multiple, select the max of the minimum required client versions.
+        /// </summary>
+        private Version GetMaxOfMinClientVersions(UserSecurityPolicyContext context)
+        {
+            var policyStates = context.Policies.Select(p => JsonConvert.DeserializeObject<State>(p.Value));
+            return policyStates.Max(s => s.MinClientVersion);
```
Comment (cristinamanum) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: Can be nulls among these values? In this case will the Max throw?

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        /// <summary>
+        /// Get the current client version from the request.
+        /// </summary>
+        private Version GetClientVersion(UserSecurityPolicyContext context)
+        {
+            var clientVersionString = context.HttpContext.Request?.Headers[Constants.ClientVersionHeaderName];
+
+            Version clientVersion;
+            return Version.TryParse(clientVersionString, out clientVersion) ? clientVersion : null;
```
Comment (cristinamanum) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: Do we want to log in AI the null case? The Evaluate below will return a non success result, the user scenario may be impacted in this case.

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        
+        public override SecurityPolicyResult Evaluate(UserSecurityPolicyContext context)
+        {
+            var minClientVersion = GetMaxOfMinClientVersions(context);
+
+            var clientVersion = GetClientVersion(context);
+            if (clientVersion == null || clientVersion < minClientVersion)
+            {
+                return new SecurityPolicyResult(false, string.Format(CultureInfo.CurrentCulture,
+                    Strings.SecurityPolicy_RequireMinClientVersionForPush, minClientVersion));
```
Comment (cristinamanum) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: Do we want to log in AI the fact that the clientVersion < minClientVersion?

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    /// <summary>
+    /// User security policy that requires a minimum client version in order to push packages.
+    /// </summary>
+    public class RequireMinClientVersionForPushPolicy : UserSecurityPolicyHandler
+    {
+        public class State
+        {
+            [JsonProperty("v")]
+            [JsonConverter(typeof(VersionConverter))]
+            public Version MinClientVersion { get; set; }
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: >Version [](start = 19, length = 7)

should be a NuGetVersion, really

File: src/NuGetGallery/Security/UserSecurityPolicyHandler.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+using NuGetGallery.Filters;
+
+namespace NuGetGallery.Security
+{
+    /// <summary>
+    /// Policy handler that defines behavior for specific user policy types.
+    /// </summary>
+    public abstract class UserSecurityPolicyHandler
+    {
+        public string Name { get; private set; }
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/UserSecurityPolicyHandler.cs]: >private [](start = 34, length = 7)

can remove private set;
This is readonly right?

File: src/NuGetGallery/Security/SecurityPolicyResult.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        public SecurityPolicyResult(bool success, string errorMessage)
+        {
+            Success = success;
+            ErrorMessage = errorMessage;
+        }
+
+        /// <summary>
+        /// Whether security policy criteria was successfully met.
+        /// </summary>
+        public bool Success { get; private set; }
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/SecurityPolicyResult.cs]: >private [](start = 35, length = 7)

can remove private set; as this is readonly

File: src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// User subscribed to this security policy.
+        /// </summary>
+        public User User { get; set; }
+
+        /// <summary>
+        /// Type name for the policy handler that provides policy behavior.
+        /// </summary>
+        [Required]
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs]: >Required [](start = 9, length = 8)

Can' this [Required] attribute be defined in EntitiesContext instead? Seems like you are requiring User.

File: src/NuGetGallery/Filters/SecurityPolicyAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+        /// <summary>
+        /// Result of the policy evaluation.
+        /// </summary>
+        public SecurityPolicyResult Result { get; private set; }
+
+        /// <summary>
+        /// Security policy service.
+        /// </summary>
+        public ISecurityPolicyService SecurityPolicyService { get; private set; }
```
Comment (joelverhagen) [inline src/NuGetGallery/Filters/SecurityPolicyAttribute.cs]: >ISecurityPolicyService [](start = 15, length = 22)

why public?

File: tests/NuGetGallery.Facts/Security/RequireMinClientVersionForPushPolicyFacts.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            // Assert
+            Assert.True(result.Success);
+            Assert.Null(result.ErrorMessage);
+        }
+
+        [Theory]
+        [InlineData("4.1.0")]
+        [InlineData("3.0.0")]
+        [InlineData("2.0.0,4.1.0")]
+        public void EvaluateReturnsFailureIfClientVersionLower(string minClientVersions)
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Security/RequireMinClientVersionForPushPolicyFacts.cs]: >EvaluateReturnsFailureIfClientVersionLower [](start = 20, length = 42)

client version is a NuGetVersion not a Version so we should test a couple NuGetVersion variants as well
- 4.1.0-beta1

**Comment by joelverhagen** [comment]
:clock1:

File: src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        public string Name { get; set; }
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs]: >Name [](start = 22, length = 4)

I think type name is an implementation detail. We should have a hard coded constant for this (either an int which is consistent with the rest of the DB or const string).

File: src/NuGetGallery/Security/RequirePackageVerifyScopePolicy.cs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace NuGetGallery.Security
+{
+    /// <summary>
+    /// User security policy to require scoped API keys with verify scope on package verification callback.
+    /// </summary>
+    public class RequirePackageVerifyScopePolicy : UserSecurityPolicyHandler
+    {
+        public RequirePackageVerifyScopePolicy()
+            : base(nameof(RequirePackageVerifyScopePolicy), SecurityPolicyAction.PackageVerify)
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/RequirePackageVerifyScopePolicy.cs]: >RequirePackageVerifyScopePolicy [](start = 26, length = 31)

Please use a const string or const int for this. nameof is easy to change with refactoring tools

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: I think we should prefer InstancePerLifetimeScope over SuntingleInstance until we have a caching/perf/behavior requirement for it. It's easier for unintended interaction between requests to occur with SingleInstance

---
In reply to: [113238685](https://github.com/NuGet/NuGetGallery/pull/3820#discussion_r113238685) [](ancestors = 113238685)

File: src/NuGetGallery/Security/SecurityPolicyResult.cs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+namespace NuGetGallery.Security
+{
+    /// <summary>
+    /// Result of a security policy evaluate action.
+    /// </summary>
+    public class SecurityPolicyResult
+    {
+        public static SecurityPolicyResult SuccessResult = new SecurityPolicyResult(true, null);
+
+        public SecurityPolicyResult(bool success, string errorMessage)
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/SecurityPolicyResult.cs]: >SecurityPolicyResult [](start = 15, length = 20)

seems like success = true and errorMessage != null is an invalid state, right? I like making the ctor private in this case and having a CreateErrorResult(string errorMessage) method.

File: src/NuGetGallery/Security/SecurityPolicyService.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+        /// <summary>
+        /// Create any supported policy handlers.
+        /// </summary>
+        protected virtual IEnumerable<UserSecurityPolicyHandler> CreateUserPolicyHandlers()
+        {
+            yield return new RequireMinClientVersionForPushPolicy();
+            yield return new RequirePackageVerifyScopePolicy();
+        }
+
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/SecurityPolicyService.cs]: nit: remove extra line

File: src/NuGetGallery/Security/RequirePackageVerifyScopePolicy.cs (hunk 15 above)
Comment (joelverhagen) [inline src/NuGetGallery/Security/RequirePackageVerifyScopePolicy.cs]: >SecurityPolicyAction [](start = 60, length = 20)

Maybe I am missing it but where are we asserting that `RequirePackageVerifyScopePolicy` has action `PackageVerify` and not something else?

**Comment by joelverhagen** [comment]
Functional test(s)?

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs (hunk 4 above)
Comment (chenriksson) [inline src/NuGetGallery.Core/Entities/EntitiesContext.cs]: Removed

File: src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs (hunk 11 above)
Comment (chenriksson) [inline src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs]: HasRequired in EntitiesContext is for a required relationship.  I added Required here since it's the only non-nullable column that isn't a PK/FK. Not sure how this actually gets used, but we don't seem to be consistent with [Required] usage.

File: src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs (hunk 14 above)
Comment (chenriksson) [inline src/NuGetGallery.Core/Entities/UserSecurityPolicy.cs]: Policy handler names are now constants. I left as strings which I think will be easier to read when querying the DB. This PR only reads policies, but the next PR (onboarding) will start writing them.

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs (hunk 6 above)
Comment (chenriksson) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: Instead of logging in individual handlers, I'm planning to add telemetry and auditing to the SecurityPolicyService in a separate PR. AI can log any policy failures and include username, policy name and value.

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs (hunk 5 above)
Comment (chenriksson) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: Shouldn't be null, but added logic to guard against it

**Comment by chenriksson** [comment]
@joelverhagen I need to create onboarded test account for the functional test. Am thinking of doing onboarding in next PR, then auditing/telemetry, then functional test.

**Comment by joelverhagen** [comment]
Please make sure the deployment task mentions running EF migrations.

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs (hunk 6 above)
Comment (cristinamanum) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: If that will be in a different PR is ok.

File: src/NuGetGallery/Filters/ApiAuthorizeAttribute.cs [hunk 18]
```diff
@@ -49,16 +74,47 @@ public override void OnAuthorization(AuthorizationContext filterContext)
                     }
                 }
             }
+            
+            // Resolve the policy service if security policy checks are required.
+            if (SecurityPolicyAction.HasValue)
+            {
+                SecurityPolicyService = ((AppController)filterContext.Controller)?.GetService<ISecurityPolicyService>();
```
Comment (loic-sharma) [inline src/NuGetGallery/Filters/ApiAuthorizeAttribute.cs]: This will throw if the `ApiAuthorizeAttribute` is used on a controller that doesn't extend `AppController`, right? I would recommend that you add documentation that this attribute __must__ be used on controllers that extend `AppController`. Also, maybe throw an exception if the controller isn't an `AppController` to make this easier to debug if this issue ever comes up?
*

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+        /// <summary>
+        /// In case of multiple, select the max of the minimum required client versions.
+        /// </summary>
+        private NuGetVersion GetMaxOfMinClientVersions(UserSecurityPolicyContext context)
+        {
+            var policyStates = context.Policies
+                .Where(p => !string.IsNullOrEmpty(p.Value))
+                .Select(p => JsonConvert.DeserializeObject<State>(p.Value));
+            return policyStates.Max(s => s.MinClientVersion);
```
Comment (loic-sharma) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: Any particular reason for the `policyStates` variable? You could merge the two statements into one and return its result directly.

File: src/NuGetGallery/Security/SecurityPolicyResult.cs [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        /// <summary>
+        /// Create a failed security policy result.
+        /// </summary>
+        /// <param name="errorMessage"></param>
+        /// <returns></returns>
+        public static SecurityPolicyResult CreateErrorResult(string errorMessage)
+        {
+            if (string.IsNullOrEmpty(errorMessage))
```
Comment (loic-sharma) [inline src/NuGetGallery/Security/SecurityPolicyResult.cs]: `string.IsNullOrEmpty` seems a little too strong. Would a null check be better here?
